Changes Nexus Tutorial to have user create a route

oatmealraisin opened the pull request:
This is better for users working on a cluster that is not on their local
machine. Also added some nits.

@bparees @ahardin-rh @openshift/team-documentation ptal?

bparees commented on dev_guide/app_tutorials/maven_tutorial.adoc:
```diff
@@ -37,29 +37,36 @@ Use `new-app` to download and deploy the official Nexus Docker image.
 ----
 
-After the build has finished, run:
+After the build has finished, create a route by exposing the newly created
```
it's not a build, just a deployment.

bparees replied:
and you don't need to wait.

bparees commented on dev_guide/app_tutorials/maven_tutorial.adoc:
```diff
@@ -86,15 +93,18 @@ $ oc set probe dc/nexus \
+xref:../volumes.adoc#dev-guide-volumes[Managing Volumes in {product-title}] for
+instructions on creating a persistent volume.
 
 Add a PVC to the Nexus deployment configuration.
```
I think @csrwng suggested you can use --replace instead of --remove and --add in the instructions below, though i'm not seeing that option in oc help right now.  Double check with him.

csrwng replied:
It's `--overwrite`

bparees replied:
not seeing that in "oc help volumes", is it a generic flag?

csrwng replied:
This is what I see. Maybe I don't have the latest?

```
$ oc set volume --help
Update volumes on a pod template

This command can add, update or remove volumes from containers for any object
that has a pod template (deployment configs, replication controllers, or pods).
You can list volumes in pod or any object that has a pod template. You can
specify a single object or multiple, and alter volumes on all containers or
just those that match a given name.

If you alter a volume setting on a deployment config, a deployment will be
triggered. Changing a replication controller will not affect running pods, and
you cannot change a pod's volumes once it has been created.

Volume types include:

* emptydir (empty directory) *default*
    A directory allocated when the pod is created on a local host, is removed when
    the pod is deleted and is not copied across servers
* hostdir (host directory)
    A directory with specific path on any host (requires elevated privileges)
* persistentvolumeclaim or pvc (persistent volume claim)
    Link the volume directory in the container to a persistent volume claim you have
    allocated by name - a persistent volume claim is a request to allocate storage.
    Note that if your claim hasn't been bound, your pods will not start.
* secret (mounted secret)
    Secret volumes mount a named secret to the provided directory.

For descriptions on other volume types, see https://docs.openshift.com

Usage:
  oc set volumes RESOURCE/NAME --add|--remove|--list [options]

Examples:
  # List volumes defined on all deployment configs in the current project
  oc set volume dc --all

  # Add a new empty dir volume to deployment config (dc) 'registry' mounted under
  # /var/lib/registry
  oc set volume dc/registry --add --mount-path=/var/lib/registry

  # Use an existing persistent volume claim (pvc) to overwrite an existing volume 'v1'
  oc set volume dc/registry --add --name=v1 -t pvc --claim-name=pvc1 --overwrite

  # Remove volume 'v1' from deployment config 'registry'
  oc set volume dc/registry --remove --name=v1

  # Create a new persistent volume claim that overwrites an existing volume 'v1'
  oc set volume dc/registry --add --name=v1 -t pvc --claim-size=1G --overwrite

  # Change the mount point for volume 'v1' to /data
  oc set volume dc/registry --add --name=v1 -m /data --overwrite

  # Modify the deployment config by removing volume mount "v1" from container "c1"
  # (and by removing the volume "v1" if no other containers have volume mounts that reference it)
  oc set volume dc/registry --remove --name=v1 --containers=c1

  # Add new volume based on a more complex volume source (Git repo, AWS EBS, GCE PD,
  # Ceph, Gluster, NFS, ISCSI, ...)
  oc set volume dc/registry --add -m /repo --source=<json-string>

Options:
      --add=false: Add volume and/or volume mounts for containers
      --all=false: select all resources in the namespace of the specified resource types
      --claim-mode='ReadWriteOnce': Set the access mode of the claim to be created. Valid values are ReadWriteOnce (rwo), ReadWriteMany (rwm), or ReadOnlyMany (rom)
      --claim-name='': Persistent volume claim name. Must be provided for persistentVolumeClaim volume type
      --claim-size='': If specified along with a persistent volume type, create a new claim with the given size in bytes. Accepts SI notation: 10, 10G, 10Gi
      --configmap-name='': Name of the persisted config map. Must be provided for configmap volume type
      --confirm=false: Confirm that you really want to remove multiple volumes
  -c, --containers='*': The names of containers in the selected pod templates to change - may use wildcards
  -f, --filename=[]: Filename, directory, or URL to file to use to edit the resource.
      --list=false: List volumes and volume mounts for containers
  -m, --mount-path='': Mount path inside the container. Optional param for --add or --remove
      --name='': Name of the volume. If empty, auto generated for add operation
  -o, --output='': Display the changed objects instead of updating them. One of: json|yaml
      --output-version='': Output the changed objects with the given version (default api-version).
      --overwrite=false: If true, replace existing volume source and/or volume mount for the given resource
      --path='': Host path. Must be provided for hostPath volume type
      --remove=false: Remove volume and/or volume mounts for containers
      --secret-name='': Name of the persisted secret. Must be provided for secret volume type
  -l, --selector='': Selector (label query) to filter on
      --source='': Details of volume source as json string. This can be used if the required volume type is not supported by --type option. (e.g.: '{"gitRepo": {"repository": <git-url>, "revision": <commit-hash>}}')
  -t, --type='': Type of the volume source for add operation. Supported options: emptyDir, hostPath, secret, configmap, persistentVolumeClaim

Use "oc options" for a list of global command-line options (applies to all commands).
```

bparees replied:
i was running "oc volumes", my bad.

bparees replied:
well actually i was doing a number of things wrong. let's not talk about it :)
(too much context switching)

ahardin-rh commented:
@oatmealraisin No new comments from me :star:

bparees commented:
@oatmealraisin this doesn't need much more work, can you wrap it up?

oatmealraisin commented:
@bparees Sorry, still running into flakeyness on changing to `--overwrite`

oatmealraisin commented:
@csrwng @bparees the issue comes from Nexus locking the working directory.  If the PV is used by subsequent pods, the deployment will fail.  I'll bring this up at tomorrow's scrum?

Otherwise, these are all of the latest changes

bparees commented:
@oatmealraisin 
hm.  the PV should be cleaned before getting reused by a different pod.  but if you mean that the existing pod restarts, i could see that being an issue.  let's talk through it tomorrow as you say.

oatmealraisin commented:
@bparees @ahardin-rh So the problem can be solved by setting up the PV with a "Recycle" policy.  I've updated one of the links with a better article on setting up PV's.

Also, I've removed the readiness probe from the example.

ptal?

bparees commented:
@oatmealraisin weird, i'd love to understand why the recycle policy is necessary for deployment to work, that should have no bearing on tearing down the ephemeral deployment.

Were you just hitting this issue on your 2nd+ run through the guide, where you'd already allocated the PV to your first deployment and then, after cleaning everything up, the PV was still claimed and thus not available to be claimed on your second attempt to create a nexus server?

anyway otherwise lgtm.  @openshift/team-documentation ptal/merge.

oatmealraisin commented:
@bparees At least from the demo, I did not clear the nexus-pv directory that I was using, so my practice beforehand doomed the demo :(

bparees commented:
@oatmealraisin yeah so if the issues you were seeing were caused by going through the steps a second time w/o sufficient cleanup beforehand, i could see that.  But if you only go through the steps a single time, I would not expect the recycle policy to matter.

vikram-redhat commented:
@bfallonf - can you please take a look?

bfallonf commented on dev_guide/app_tutorials/maven_tutorial.adoc:
```diff
@@ -37,29 +37,35 @@ Use `new-app` to download and deploy the official Nexus Docker image.
-Confirm that Nexus is running by navigating your browser to the new Cluster-IP
-at port 8081. The server may take a bit of time to start, be patient!
+Confirm that Nexus is running by navigating your browser to the URL under
+*HOST/PORT*. To sign into Nexus, the default administrator username is *admin*,
```
s/sign into/sign in to

bfallonf commented on dev_guide/app_tutorials/maven_tutorial.adoc:
```diff
@@ -86,29 +92,32 @@ $ oc set probe dc/nexus \
-Read/Write Persistent Volume for you.  Otherwise, read
-xref:../volumes.adoc#dev-guide-volumes[Managing Volumes in {product-title}] for instructions on
-creating a persistent volume.
+will not be lost when the pod running the server terminates. PVCs require
```
s/will not be lost/are not lost

bfallonf commented on dev_guide/app_tutorials/maven_tutorial.adoc:
```diff
@@ -86,29 +92,32 @@ $ oc set probe dc/nexus \
+available persistent volumes (PV) in your cluster. If there are no PVs
+available in your cluster and you do not have administrator access on your
+cluster, ask your system administrator to create a Read/Write Persistent Volume
+for you.  Otherwise, read
```
s/read/see

bfallonf commented:
@vikram-redhat @oatmealraisin A few sentence-level stuff. LGTM otherwise!

oatmealraisin commented:
@bfallonf Done :)

php-coder commented on dev_guide/app_tutorials/maven_tutorial.adoc:
```diff
@@ -86,29 +87,32 @@ $ oc set probe dc/nexus \
+persistent volumes (PV) in your cluster. If there are no PVs available in your
+cluster and you do not have administrator access on your cluster, ask your
+system administrator to create a Read/Write Persistent Volume for you.
+Otherwise, see
```
This paragraph has too much occurrences of `your` word. Maybe *you* can improve it? :-)

oatmealraisin replied:
@php-coder How about now?

php-coder replied:
It's better now, but WDYT about s/your system administrator/a system administrator/ ?

oatmealraisin replied:
They are definitely different tones, but I don't really mind either in this place.  "your" sounds more concrete as opposed to "a" which could be anyone.. maybe I'm overthinking it.  How about you?

php-coder replied:
I don't mind to leave it as is :)

oatmealraisin commented:
@openshift/team-documentation ptal?

vikram-redhat commented:
@bfallonf - please take a look at the updates.

bfallonf commented:
@vikram-redhat @oatmealraisin LGTM!

bparees commented:
Merge?

ahardin-rh commented:
No rev history needed, as it is covered in https://github.com/openshift/openshift-docs/pull/2778
